### ISSUE TITLE
When using allow_scheduling_on_control_plane=true, lb does not point to control planes

### ISSUE THREAD

**Issue by leonsteinhaeuser** [kube-hetzner/terraform-hcloud-kube-hetzner#55]
**What I did:**

```bash
git clone https://github.com/kube-hetzner/kube-hetzner.git
git checkout staging

export TF_VAR_hcloud_token="my-hcloud-token"
```

Create a terraform.tfvars file:
```hcl
# You need to replace these
public_key   = "~/.ssh/id_ed25519.pub"
# Must be "private_key = null" when you want to use ssh-agent, for a Yubikey like device auth or an SSH key-pair with passphrase
private_key  = "~/.ssh/id_ed25519"

# These can be customized, or left with the default values
# For Hetzner locations see https://docs.hetzner.com/general/others/data-centers-and-connection/
# For Hetzner server types see https://www.hetzner.com/cloud
location                  = "nbg1" # change to `ash` for us-east Ashburn, Virginia location
network_region            = "eu-central" # change to `us-east` if location is ash
agent_server_type         = "cpx21"
control_plane_server_type = "cpx31"
lb_server_type            = "lb11"
servers_num               = 3
agents_num                = 0

# If you want to use a specific Hetzner CCM and CSI version, set them below, otherwise leave as is for the latest versions
# hetzner_ccm_version = ""
# hetzner_csi_version = ""

# If you want to kustomize the Hetzner CCM and CSI containers with the "latest" tags and imagePullPolicy Always, 
# to have them automatically update when the node themselve get updated via the rancher system upgrade controller, the default is "false".
# If you choose to keep the default of "false", you can always use ArgoCD to monitor the CSI and CCM manifest for new releases,
# that is probably the more "vanilla" option to keep these components always updated. 
# hetzner_ccm_containers_latest = true
# hetzner_csi_containers_latest = true

# If you want to use letsencrypt with tls Challenge, the email address is used to send you certificates expiration notices
traefik_acme_tls = true
traefik_acme_email = "my@mail.com"

# If you want to allow non-control-plane workloads to run on the control-plane nodes set "true" below. The default is "false".
allow_scheduling_on_control_plane = true
```

**What happened:**

All resources are properly scheduled, but the load balancer does not point to the control planes.

![lb_overview](https://user-images.githubusercontent.com/42437185/153226525-16e8510c-7922-4063-8145-1b4d65f34293.jpg)

**What I expect:**

I expect a reference from the load balancer to the control planes.

**Comment by mysticaltech** [comment]
Yes, this is normal, because it will show up if you create an ingress definition to expose a service of a sort. So it will work.

**Comment by mysticaltech** [comment]
Ok thanks, will investigate further!

**Comment by mysticaltech** [comment]
Confirmed, my control planes have:
```
Taints:             <none>
Unschedulable:      false
```

Yet, the LB only targets the agents, as follows, obtained with `hcloud load-balancer describe traefik`:

![ksnip_20220210-002114](https://user-images.githubusercontent.com/518555/153307456-51a2f06d-158e-43a3-8e8c-68deb9286584.png)

**Comment by mysticaltech** [comment]
@leonsteinhaeuser Please have a look at https://github.com/traefik/traefik-helm-chart/blob/master/traefik/values.yaml, and search the web about the matter, as my initial attempt at finding a solution failed.

If that can be set via the traefik config helm, then we can do so via the `traefik_config.yaml.tpl `HelfChartConfig  file.

**Comment by mysticaltech** [comment]
I believe this has to do with https://github.com/hetznercloud/hcloud-cloud-controller-manager, which does not seem to want to configure the ingress controller with control plain nodes IPs, even if those have scheduling enabled. 

So, unfortunately, there is nothing we can do on our side but it would be wonderful if you could open an issue here https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues @leonsteinhaeuser.

**Comment by geraldwuhoo** [comment]
It looks like in https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/101#issuecomment-716102389, the reason is actually because k8s itself reports the list of nodes to the CCM, and it will not report the controlplane nodes as schedulable as long as they are tainted *and* labeled as master nodes. The response seems to imply that removing the taint *and* label from the controlplane nodes fixed the issue.

For this project, I believe removing the taints and labels from all controlplane nodes when controlplanes are allowed schedulable is a reasonable action. What do you think?

**Comment by mysticaltech** [comment]
@geraldwuhoo Please, if you can, try issuing `kubectl label nodes --all node-role.kubernetes.io/master-` and let me know if that works! 🙏

If it does - PR is welcome! :)

**Comment by mysticaltech** [comment]
@leonsteinhaeuser, Please also give that a shot if you are still using the cluster! Between, you might want to redeploy as it has improved light years since February! :)

**Comment by geraldwuhoo** [comment]
> @geraldwuhoo Please, if you can, try issuing `kubectl label nodes --all node-role.kubernetes.io/master-` and let me know if that works! pray
> 
> If it does - PR is welcome! :)

Just tried this on a fresh cluster &mdash; I can confirm that my controlplane nodes show up under the loadbalancer when I remove the master label from them.

**Comment by geraldwuhoo** [comment]
Update on this: removing the labels affects the k3s auto-upgrade ability, as the upgrade script uses the master label to select the controlplane nodes for upgrade. In any case, it looks like removing the master label is a bit brittle anyway: https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/197. However, this issue claims that passing the `LegacyNodeRoleBehavior=false` feature gate flag in the creation of the hcloud CCM allows for scheduling on controlplane nodes. Should be easy enough to add to this file: 

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/20a6ebc667d535d2dbb81c1f9983219df97829f7/templates/ccm.yaml.tpl

I don't think this will adversely affect existing deployments, as this label has apparently already been locked to false for a while, and is fully removed in v1.22. Or perhaps it could be conditionally set only when `allow_scheduling_on_control_plane` is true.

**Comment by mysticaltech** [comment]
Great to hear @geraldwuhoo! Please, don't hesitate to create a PR to the staging branch. Otherwise I will work on it whenever I have time. Great work!!

**Comment by mysticaltech** [comment]
@leonsteinhaeuser This is now taken care of by the PR above! Will be shipped in the next release! @geraldwuhoo! 🥳

**Comment by mysticaltech** [comment]
Released in v1.2.4!
